python/ray/serve/_private/controller.py

```diff
@@ -1032,6 +1046,63 @@ def _get_logging_config(self) -> Tuple:
         return self.system_logging_config, log_file_path
 
 
+def calculate_scale_direction(
+    new_config: ServeDeploySchema,
+    all_app_statuses: Dict[str, ApplicationStatusInfo],
+    curr_target_capacity: Optional[float],
+    curr_scale_direction: Optional[float],
+) -> Optional[TargetCapacityScaleDirection]:
+    next_scale_dirction = None
+
+    if len(all_app_statuses) == 0 and new_config.target_capacity is not None:
+        next_scale_dirction = TargetCapacityScaleDirection.UP
+
+    if live_applications_match_config(new_config, all_app_statuses):
+        next_target_capacity = new_config.target_capacity
+
+        if curr_target_capacity == next_target_capacity:
+            next_scale_dirction = curr_scale_direction
+        elif curr_target_capacity is None and next_target_capacity is not None:
+            next_scale_dirction = TargetCapacityScaleDirection.DOWN
```

Comment:
If `curr_target_capacity is None` and if there are applications running, then the config has finished deploying, and Serve is rolled out `100%` and in steady-state.

We have 2 options for how to handle the case where `next_target_capacity is not None` but the apps in the config match the apps in the previous config:

1. We could consider this a rollback, since the new config is the "same" as the previous config, except with a lower `target_capacity`, so we set the direction to `TargetCapacityScaleDirection.DOWN`.
2. We could consider this a roll forward to an entirely new version of the "same" config, since the current config has totally finished rolling out. In that case, we set the direction to `TargetCapacityScaleDirection.UP`.

I went with (1), but I'm curious what you think. Do you think (2) makes more sense?

Reply:
I think it all boil down to what we want the lower bound to be. I'm fine with either option, but would be nice to add a detailed comment like you just mentioned here in the code :)

Reply:
I thought about it some more. I think (2) actually makes more sense. Once `curr_target_capacity is None`, we should consider Serve to be in steady-state. When we apply a new config, we should consider that a roll forward of an entirely new version, rather than comparing it to the previous version which has already rolled out.

Reply:
I added a comment.